028-service-binding.md

```diff
@@ -0,0 +1,340 @@
+ - Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+
+Considerations:
+
+ - In this approach a lot of work is still left up to the user to discover the cluster certificate and the bootstrap server address to use.
+
+### 2 - Application uses multiple ServiceBinding resources
+
+In this approach the application creates two `ServiceBinding` resources, one that will bind to the `Kafka` CR and one that will bind to the `KafkaUser` CR.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the `Kafka` CR to include a `status.binding` field, where the `name` refers to a new secret containing both the cluster certificate and bootstrapServers
+- Add a new secret that contains the bootstrap servers for each listener and the cluster certificate information
+
+Considerations:
+
+ - This approach results in a new secret being required that contains copies of existing secrets
+ - It does not help the user to decide which listener to use
+
+### 3 - Application binds to KafkaUser resource that contains all details
+
+In this approach the application only binds to the `KafkaUser` and the secret created for the `KafkaUser` contains all the required details.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the secret created for the `KafkaUser` to include the bootstrap servers for each listener
+- Update the secret created for the `KafkaUser` to include the cluster certificate information
+
+Considerations:
+
+- This approach results in a lot more information being added to each `KafkaUser` secret
+- It does not help the user to decide which listener to use
+
+Possible extensions:
+
+ - Could only add bootstrap server addresses for listeners that support the authentication type that the `KafkaUser` supports
+
+### 4 - Application can bind to a single custom resource
+
+In this approach the application binds to a new custom resource type, e.g. called `KafkaConnection`.
+
+Required changes:
+
+ - Add a new Custom Resource called `KafkaConnection`
+ - The `KafkaConnection` spec has several fields, cluster, listener and user, that refer to the `Kafka`, the specific listener and the `KafkaUser` that should be used
+ - The Strimzi operator adds a `spec.binding` to the `KafkaConnection` status
+ - The Strimzi operator add a new secret containing all the required information
+ 
+Considerations:
+
+- This approach results in a new Custom Resource type that is just for convenience
```

Comment:
I would question how convenient it really is over the `KafkaUser` CR supporting a `spec.binding.listener` idea. It requires extra CRs, and keeping the references straight. It doesn't really reduce the coupling to the service binding spec (i.e. we'd have CRs which had such a coupling in either case).

Reply:
If the user has authorization turned off, then they wouldn't normally have to create a KafkaUser, or use the entity operator, but to make this work they would. I think the actual advantage of this is where Jakub was suggesting that this new CR could live in a different namespace to the install of Strimzi. Is it possible to put your KafkaUser CR in a different namespace to the Strimzi instance it refers to?

Reply:
The `KafkaUser` operator can watch a different namespace than the one it's deployed in, but the `Secrets` will currently always be in the same namespace as the `KafkaUser` CR. But there's nothing stopping us from allowing the `KafkaUser` spec to nominate a (set of) namespace(s) to publish service-binding compatible `Secrets` to. Or doing that at the operator level (e.g. add `PUBLISH_NAMESPACE` alongside `WATCHED_NAMESACE`).

I'm not necessarily opposed to a new CR, but I'd like to understand why that's better than using an improved `KafkaUser`.

Reply:
> The `KafkaUser` operator can watch a different namespace than the one it's deployed in, but the `Secrets` will currently always be in the same namespace as the `KafkaUser` CR. 

The typical use-case is that you have application per-namespace. So for 99% of cases, the _watch a different namespace_ feature is useless since it changes the namespace, but still keeps everything in one namespace.

> But there's nothing stopping us from allowing the `KafkaUser` spec to nominate a (set of) namespace(s) to publish service-binding compatible `Secrets` to. Or doing that at the operator level (e.g. add `PUBLISH_NAMESPACE` alongside `WATCHED_NAMESACE`).

I think one of the challenges is that you cannot use garbage collection, so we will be leaving the secrets behind. (dangerous with TLS where the certificate works even after the user is deleted). It is also a bit messy from the RBAC perspective. it means that the CO will need to have the rights this `strimzi-binding` operator might need. Users might be also asking for a weird things => such as "I want to give it access to secrets in these 10 namespaces but not whole cluster. If CO does the RBAC setup for UO, this will be hard to do. 

So I feel like separation of concerns makes better sense. And it solves the namespace issue for us as well to some extent. 

The advantage of using KafkaUser would be especially if we one day really implement multi-namespace support for the UO.

Reply:
> If the user has authorization turned off, then they wouldn't normally have to create a KafkaUser, or use the entity operator, but to make this work they would.

I think this is the bigger point about not using the `KafkaUser`. With solution 4, the `KafkaConnection` would be anyway created with no auth data but just the listener ones and it would make more sense to me.

Reply:
> I think the actual advantage of this is where Jakub was suggesting that this new CR could live in a different namespace to the install of Strimzi.

I would see that as a big plus, because application needs to get access only to KafkaConnection living in the same namespace as the app. Speaking about garbage collection - by l8s there are no ownership relationship across namespaces, but it can be tracked by Kafka and User operator, so whenever of them gets removed, `KafkaConnection` can be cleaned up as well.